diff --git a/fs/Kconfig b/fs/Kconfig
index cf6ae64..e358462 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -130,6 +130,7 @@ source "fs/quota/Kconfig"
 source "fs/autofs/Kconfig"
 source "fs/fuse/Kconfig"
 source "fs/overlayfs/Kconfig"
+source "fs/hangar/Kconfig"
 
 menu "Caches"
 
diff --git a/fs/Makefile b/fs/Makefile
index 89a8a9d..1e4e757 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -103,6 +103,7 @@ obj-$(CONFIG_AUTOFS_FS)		+= autofs/
 obj-$(CONFIG_ADFS_FS)		+= adfs/
 obj-$(CONFIG_FUSE_FS)		+= fuse/
 obj-$(CONFIG_OVERLAY_FS)	+= overlayfs/
+obj-$(CONFIG_HANGAR_SYNC)	+= hangar/
 obj-$(CONFIG_ORANGEFS_FS)       += orangefs/
 obj-$(CONFIG_UDF_FS)		+= udf/
 obj-$(CONFIG_SUN_OPENPROMFS)	+= openpromfs/
diff --git a/fs/locks.c b/fs/locks.c
index 6e4ff7f..587313f 100644
--- a/fs/locks.c
+++ b/fs/locks.c
@@ -52,6 +52,7 @@
 #include <linux/file.h>
 #include <linux/fdtable.h>
 #include <linux/filelock.h>
+#include <linux/hangar_sync.h>
 #include <linux/fs.h>
 #include <linux/init.h>
 #include <linux/security.h>
@@ -2313,6 +2314,9 @@ SYSCALL_DEFINE2(flock, unsigned int, fd, unsigned int, cmd)
 		return error;
 
 	can_sleep = !(cmd & LOCK_NB);
+	error = hangar_sync_lock(fd_file(f), &fl, can_sleep);
+	if (error)
+		return error;
 	if (can_sleep)
 		fl.c.flc_flags |= FL_SLEEP;
 
@@ -2323,6 +2327,7 @@ SYSCALL_DEFINE2(flock, unsigned int, fd, unsigned int, cmd)
 	else
 		error = locks_lock_file_wait(fd_file(f), &fl);
 
+	hangar_sync_changed(fd_file(f));
 	locks_release_private(&fl);
 	return error;
 }
@@ -2529,6 +2534,10 @@ static int do_lock_file_wait(struct file *filp, unsigned int cmd,
 	if (error)
 		return error;
 
+	error = hangar_sync_lock(filp, fl, cmd == F_SETLKW);
+	if (error)
+		return error;
+
 	for (;;) {
 		error = vfs_lock_file(filp, cmd, fl, NULL);
 		if (error != FILE_LOCK_DEFERRED)
@@ -2539,6 +2548,7 @@ static int do_lock_file_wait(struct file *filp, unsigned int cmd,
 			break;
 	}
 	locks_delete_block(fl);
+	hangar_sync_changed(filp);
 
 	return error;
 }
@@ -2797,6 +2807,7 @@ void locks_remove_posix(struct file *filp, fl_owner_t owner)
 	if (lock.fl_ops && lock.fl_ops->fl_release_private)
 		lock.fl_ops->fl_release_private(&lock);
 	trace_locks_remove_posix(inode, &lock, error);
+	hangar_sync_changed(filp);
 }
 EXPORT_SYMBOL(locks_remove_posix);
 
@@ -2868,6 +2879,7 @@ void locks_remove_file(struct file *filp)
 	locks_check_ctx_file_list(filp, &ctx->flc_flock, "FLOCK");
 	locks_check_ctx_file_list(filp, &ctx->flc_lease, "LEASE");
 	spin_unlock(&ctx->flc_lock);
+	hangar_sync_changed(filp);
 }
 
 /**
